source/mir/combinatorics/package.d

```diff
@@ -613,14 +718,59 @@ pure nothrow @safe unittest
     assert(d.length == 1);
 }
 
+static if (__VERSION__ > 2069) unittest
+{
+    // check invalid
+    import std.exception: assertThrown;
+    import core.exception: AssertError;
+    import std.experimental.allocator.mallocator : Mallocator;
+
+    assertThrown!AssertError(0.cartesianPower(0));
+    assertThrown!AssertError(Mallocator.instance.makeCartesianPower(0, 0));
+}
+
+// length
+pure nothrow @safe unittest
+{
+    assert(1.cartesianPower(1).length == 1);
+    assert(1.cartesianPower(2).length == 1);
+    assert(2.cartesianPower(1).length == 2);
+    assert(2.cartesianPower(2).length == 4);
+    assert(2.cartesianPower(3).length == 8);
+    assert(3.cartesianPower(1).length == 3);
+    assert(3.cartesianPower(2).length == 9);
+    assert(3.cartesianPower(3).length == 27);
+    assert(3.cartesianPower(4).length == 81);
+    assert(4.cartesianPower(10).length == 1048576);
+    assert(14.cartesianPower(7).length == 105413504);
+}
+
 /**
-Lazily computes all k-combinations of $(D r).
+Disposes a CartesianPower object. It destroys and then deallocates the
+CartesianPower object pointed to by a pointer.
+It is assumed the respective entities had been allocated with the same allocator.
+
+Params:
+    alloc = Custom allocator
+    perm = CartesianPower object
+
+See_Also:
+    $(LREF makeCartesianPower)
+*/
+void dispose(Allocator)(auto ref Allocator alloc, auto ref CartesianPower cartesianPower)
+{
+    import std.experimental.allocator: dispose;
+    dispose(alloc, cartesianPower._state);
+}
+
+/**
+Lazily computes all k-combinations of `r`.
 Imagine this as the $(LREF cartesianPower) filtered for only strictly ordered items.
 
 While generating a new combination is in `O(k)`,
 the number of combinations is `binomial(n, k)`.
 
-For example $(D"AB".combinations(2).array) returns $(D["AB"]).
+For example `"AB".combinations(2).array` returns `["AB"]`.
```

Comment:
remove this example from the docs. And add constraints, that templates are note accepts narrow strings.